doc/compiler/osr/OSR.md

```diff
@@ -0,0 +1,275 @@
+
+**Fear point analysis**: is a backward union analysis that determines whether it's possible for execution to reach a fear point without first hitting an OSR yield point. OSR guards are inserted after each OSR yield point at which this is possible. This ensures that every fear point is protected by the closest possible OSR guard. In addition, all of the resulting OSR guards are in positions where it's possible to induce OSR because HCR guard analysis has prevented the creation of any fear points that would require OSR guards where OSR is uninducible.
+
+OSR guards are placed after OSR yield points and behave similarly to HCR guards. The guard will not necessarily be placed immediately after the OSR yield point,
+rather several stores may be located between the two. These will include stores to pending push slots, to determine the stack at the transition point (as these are used for rematerialization),
+and stores of call or allocation results. The fall through path will be a path to a fear point without going through any other OSR guards,
+but the taken path will point to an **OSR induce block**. OSR induce block will contain call to the **jitInduceOSRAtCurrentPC** helper and will have exception edges to the **OSR catch block**.
+
+As an optimization, we might have a store of the call result into an auto other than a pending push temp. This is because call results are often immediately stored to autos in the bytecode. Since we need to store it to the auto anyway, we want to avoid storing to a pending push temp, then potentially inducing OSR, and then storing again to the original auto. Instead we notice that the call result is stored into an auto (and presumably that what we store into the auto is always the result of the call, i.e. there's no branch target in between), and we push the OSR point forward past the store.
+
+Runtime assumptions are added to patch the inserted OSR guards in the event of a method redefinition,
+triggering the transition of execution to the JVM via the OSR mechanism. The OSR guards are permanently patched during an OSR yield to the JVM at which invalidation happens (e.g. a redefinition occurs). An important distinction is that all OSR guards will be patched in the event of any redefinition,
+rather than just the required set to protect a removed HCR guard. This is due to the high overhead of tracking and describing the minimal set of OSR guards that would need to be patched on invalidation of each individual assumption.
+
+## OSR and virtual guard merging
+
+As an optimization, we can mark some existing virtual guard as an OSR guard, therefore merging them together. This is achieved by making virtual guards kill fear. In this case, the taken side of a merged OSR guard will be a cold call (vs. OSR induce block). The disadvantage of the guard merging is that the merged guard cannot be removed anymore, while the original virtual guard could've been removed as a result of later optimizaitons.
```

Comment:
*optimizaitons* → *optimizations*

(though I've suggested a change that would avoid this word here if you choose to use it or something similar)